src/lib/storage/table.cpp

```diff
@@ -1,84 +1,111 @@
 #include "table.hpp"
-
+#include "resolve_type.hpp"
 #include "utils/assert.hpp"
 
 namespace opossum {
 
 Table::Table(const ChunkOffset target_chunk_size) {
-  // Implementation goes here
+  _target_chunk_size = target_chunk_size;
+  _chunks = std::vector<std::shared_ptr<Chunk>>{std::make_shared<Chunk>()};
 }
 
 void Table::add_column_definition(const std::string& name, const std::string& type, const bool nullable) {
-  // Implementation goes here
+  _column_names.emplace_back(name);
+  _column_types.emplace_back(type);
+  _column_nullable.emplace_back(nullable);
 }
 
 void Table::add_column(const std::string& name, const std::string& type, const bool nullable) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  DebugAssert(_chunks[0]->size() == 0, "Adding columns is only allowed if the table does not have any entries.");
+  add_column_definition(name, type, nullable);
+  resolve_data_type(type, [&](const auto data_type_t) {
+    using ColumnDataType = typename decltype(data_type_t)::type;
+    const auto value_segment = std::make_shared<ValueSegment<ColumnDataType>>(nullable);
+    _chunks[0]->add_segment(value_segment);
+  });
 }
 
 void Table::create_new_chunk() {
-  // Implementation goes here
+  _chunks.emplace_back(std::make_shared<Chunk>());
+
+  for (unsigned int i = 0; i < _column_types.size(); i++) {
+    resolve_data_type(_column_types[i], [&](const auto data_type_t) {
+      using ColumnDataType = typename decltype(data_type_t)::type;
+      const auto value_segment = std::make_shared<ValueSegment<ColumnDataType>>(_column_nullable[i]);
+      _chunks.back()->add_segment(value_segment);
+    });
+  }
 }
 
 void Table::append(const std::vector<AllTypeVariant>& values) {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  if (_chunks.back()->size() == _target_chunk_size) {
+    create_new_chunk();
+  }
+  _chunks.back()->append(values);
 }
 
 ColumnCount Table::column_count() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return static_cast<ColumnCount>(_column_names.size());
 }
 
 uint64_t Table::row_count() const {
-  // Implementation goes here
-  Fail("Implementation is missing.");
+  return (_chunks.size() - 1) * _target_chunk_size + _chunks.back()->size();
```

Comment:
This works, when we are going to have immutable chunks only, but I am not sure about this. Probably a more simple approach having less assumptions:

```
auto sum = uint32_t{0};
for (const auto& chunk : _chunks) {
  sum += chunk.size();
}
```

Reply:
The exercise states that 

"When a table is created, an optioanle parameter defines the maximum size of a chunk. [...]. The maximum chunk size is stored in the table and cannot be changed. Inserts are always done into the last chunk, checking if this chunk has already reached its maximum size. If this is the case, a new chunk is created." 

What do you think? Should we change the implementation or keep it?